man/man1/xapers.1
.\" xapers - journal article indexing system
.\"
.\" Copyright © 2013 Jameson Rollins
.\"
.\" Xapers is free software: you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation, either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" Xapers is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program.  If not, see http://www.gnu.org/licenses/ .
.\"
.\" Author: Jameson Rollins <jrollins@finestructure.net>
.TH XAPERS 1
.SH NAME
xapers \- personal journal article indexing system
.
.SH SYNOPSIS
.B xapers
.IR command " [" args " ...]"
.
.SH DESCRIPTION
Xapers is a personal document indexing system, geared towards academic
journal articles.  It provides fast search of document text and
bibliographic data (synced from online libraries) and simple document
and bibtex retrieval.

Xapers takes as input document files (as PDF) and source identifiers.
Documents are copied into a local document store (~/.xapers/docs by
default) and text is extracted from the PDF and fully indexed into a
Xapian database.  Source identifiers are used to download document
bibliographic data from online digital libraries (see \fBSOURCES\fR
below), which are then parsed and indexed to prefixed terms in the
database.  The bibliographic data is also stored as bibtex in the
document store for easy retrieval.  Documents can be arbitrarily
tagged.

A curses UI is provided for simple access to documents (see the
\fBview\fR command below).  Xapers is ultimately a document indexing
library, though, so development of alternate user interfaces is
encouraged.

Underlying Xapers is the wonderful Xapian database/search engine.  See
http://xapian.org/ for more information.
.
.SH MAIN COMMANDS

The following are the main xapers commands.  See \fBSEARCH TERMS\fR
below for details of the supported syntax for <search-terms>.
.
.SS add [options] [<search-terms>]

Add a document, or update an existing document.  Must specify at least
one of \-\-file or \-\-source.  If search terms are provided they must
match exactly one document and the matching document is updated with
the newly provided information.  Available options:
.RS 4
.TP 4
.BR \-\-source=[<sid>|<file>]
Source identifier for document.  See \fBSOURCES\fR below.  This may
also be a path to a file that contains a single bibtex entry.
.RE
.RS 4
.TP 4
.BR \-\-file[=<file>]
Document file (as PDF) to add.  Text of document will be extracted and
indexed.  A copy of the file will be placed in the Xapers document
store.  If provided without path, xapers will attempt to download file
from source, assuming source supports file downloads.
.RE
.RS 4
.TP 4
.BR \-\-tags=<tag>[,...]
Initial tags to apply to document.  Multiple tags can be specified,
comma separated.
.RE
.RS 4
.TP 4
.BR \-\-prompt
Prompt user for source/file/tags, if not specified.  When prompting
for source information input files are automatically scanned for
source IDs and found ids are displayed.
.RE
.RS 4
.TP 4
.BR \-\-view
View resulting entry in curses UI when done.  See the \fBviewP\fR
command below for more info.
.RE
.
.SS import [options] <bibtex>

Import an existing bibtex database.  Each bibtex entry will be added
as a new document.  If bibtex key, or any sources found in bibtex,
match an existing document, that document is instead updated (this
makes the command effectively idempotent).  Any "file" fields will be
parsed for document files to add.  Files can be specified as a single
path, or in Mendeley/Jabref format.  Available options:
.RS 4
.TP 4
.BR \-\-tags=<tag>[,...]
Tags to apply to all imported documents.  Multiple tags can be
specified, comma separated.
.RE
.
.SS tag +<tag>|-<tag> [...] [--] <search-terms>

Add/remove tags from documents.  '--' can be used to separate tagging
operations from search terms.
.
.SS search [options] <search-terms>

Search for documents in the database.  Document information is printed
to stdout.
.RS 4
.TP 4
.BR \-\-output=[summary|bibtex|tags|sources|keys|files]

Specify document information to be output:

.B summary
outputs a single-line summary of the documents (default).

.B bibtex
outputs bibtex for all documents (if available).

.B tags
outputs all tags associated with documents.

.B sources
outputs all sources associated with documents.

.B keys
outputs all bibtex citation keys associated with documents.

.B files
outputs the full paths to all files associated with documents.

Default is
.B summary.
.RE
.RS 4
.TP 4
.BR \-\-limit=N

Limit number of results returned to N.
.RE
.
.SS bibtex <search-terms>

Short for "search \-\-output=bibtex <search-terms>".
.
.SS count <search-terms>

Return a simple count of search results.
.
.SS view [<search-terms>]
.SS show [<search-terms>]

View search results in curses search UI.  Documents matching search
are displayed with their bibliographic information and a short text
summary.  It allows for manipulating document tags and for retrieved
for document files and source URLs for viewing (see
.B xdg-open(1)
for more info).  Initial search terms can be provided, but further
searches can be performed from within the UI.  While in the UI type
"?" for available commands.

NOTE: At the moment only the top 20 search results are displayed, due
to synchronous loading restrictions.  This obviously needs to be
fixed.
.
.SS export <directory> <search-terms>

Copy PDF files of resulting documents into <directory>, named with
document titles when available.
.
.SS delete <search-terms>

Delete documents from the database.  All document files will purged
from the document store.
.RS 4
.TP 4
.BR \-\-noprompt
Do not prompt to confirm deletion of documents.
.RE
.
.SS restore

Restore a database from existing xapers root.
.
.SH SOURCE COMMANDS

These commands provide access to some of the source module methods.  See
\fBSOURCES\fR below.
.
.SS sources

List available sources.
.
.SS source2url <sid> [<sid>...]

Parse a source identifier string and print the corresponding source
URL.
.
.SS source2bib <sid> [<sid>...]

Retrieve bibtex from source for a specified URL or source id, and
write to stdout.
.
.SS source2file <sid>

Retrieve file from source for a specified URL or source id, and write
to stdout.
.
.SS scandoc <file>

Scan a document file (PDF) for source IDs, and print and recognized
source ids to stdout.
.
.SH SOURCES

Sources are online databases from which document bibliographic data
can be retrieved.  In Xapers, online libraries are assigned unique
prefixes.  The online libraries associate unique document identifiers
to individual documents.  See 'xapers sources' for a list of available
online sources.

Xapers recognizes document a source identifier, or \fBsid\fR, in two
forms:

  full URL                      http://dx.doi.org/10.1364/JOSAA.29.002092
  sid of form <source>:<id>     doi:10.1364/JOSAA.29.002092

URLs are parsed into sources and source ids when recognized, and this
information is used to retrieve bibtex from the online library
databases.  The sources and sids for a given document are stored as
prefixed terms in the Xapers database (see below).
.
.SH SEARCH TERMS

Xapers supports a common syntax for search terms.

Search can consist of free-form text and quoted phrases.  Terms can be
combined with standard Boolean operators.  All terms are combined with
a logical OR by default.  Parentheses can be used to group operators,
but must be protect from shell interpretation.  The string '*' will
match all documents in the database.

Additionally, the following prefixed terms are understood (where
<brackets> indicate user-supplied values):

    id:<docid>                   Xapers document ID
    author:<string>              string in authors (also a:)
    title:<string>               string in title (also t:)
    tag:<tag>                    specific user tag
    <source>:<id>                specific source id (sid)
    source:<source>              specific source
    key:<key>                    specific bibtex citation key
    year:<year>                  specific publication year (also y:)
    year:<since>..<until>        publication year range (also y:)
    year:..<until>
    year:<since>..

Publication years must be four-digit integers.

See the following for more information on search terms:

  http://xapian.org/docs/queryparser.html
.
.SH ENVIRONMENT
The following environment variables can be used to control the
behavior of xapers:
.
.SS XAPERS_ROOT
Location of the Xapers document store.  Defaults to "~/.xapers/docs"
if not specified.
.
.SS XAPERS_SOURCE_PATH
Path specification for location of additional custom Xapers source
modules.  Defaults to "~/.xapers/sources" if not specified.
.
.SH CONTACT
Feel free to email the author:

    Jameson Rollins <jrollins@finestructure.net>


NEWS
Xapers 0.6 (2015-01-10)
=======================

* lib: Document year is now indexed as a proper value, supporting both
  'year:<year>' binary and 'year:<year>..<year>' range searches.  Will
  require running 'restore' from the cli to re-index previously added
  articles.  Thanks to Rafael Laboissière for the contribution.

* lib: Much improved performance of Document prefixed term retrieval.

* cli: Much improved search output performance.  Default output limit
  of 20 has been removed.

* nci: Curses interface now loads entries dynamically, so search
  results are no longer limited.

* nci: Cleaner results display.  Document 'summary' has been removed,
  but other information has been added, such as journal article, year,
  file info, and document match information.

* nci: Extended navigation commands.

* sources: Now support PDF download from sources, if the source
  supports it.

* sources: New IACR 'crytoeprint' source.  Thanks to Daniel Kahn
  Gillmor for the contribution.

* Other cleanup, bug fixes, and performance improvements.


xapers.py
#!/usr/bin/env python

import os
import sys
import xapers

########################################################################

# combine a list of terms with spaces between, so that simple queries
# don't have to be quoted at the shell level.
def make_query_string(terms):
    return str.join(' ', terms)

########################################################################

def usage():
    prog = os.path.basename(sys.argv[0])
    print "Usage:", prog, "<command> [args...]"
    print """
  add [options]                               add new document
    --source=source                             specify source
    --file=file                                 file to index
    --tags=tag[,...]                            initial tags
    --prompt                                    prompt for unspecified options
    --view                                      view entry after adding
  update [options] docid                      update document
    --source=source                             specify source
    --file=file                                 file to index
  delete docid                                delete document from database

  tag +tag|-tab [...] [--] search-terms...    add/remove tags

  search [options] search-terms...            search the database
    --output=[files|summary|bibtex|sources|tags] output format
    --limit=N                                    limit number returned (20)
  bibtex search-terms...                      search --output=bibtex
  view search-terms...                        view search in selector UI
  count search-terms...                       count matches

  export dir search-terms...                  export documents to a directory,
                                              docs named based on contents

  source2bib source                           retrieve bibtex for source

  version
  help                                        this help

sources: sources can be either urls, or 'source:id' strings, or bibtex files.
search-terms: terms to search for, included prefixed terms:
  id:        document id
  source:    source type
  bib:       bibtex db entry key
  author:    author string
  title:     title string
  tag:       user tags
"""

if __name__ == '__main__':

    try:
        xdir = os.environ['XAPERS_DIR']
    except:
        print >>sys.stderr, "XAPERS_DIR environment variable not specified."
        sys.exit(1)
    if not os.path.isdir(xdir):
        print >>sys.stderr, "XAPERS_DIR '%s' does not exist." % (xdir)
        sys.exit(2)
    if 'XAPERS_DB' in os.environ:
        xdb = os.environ['XAPERS_DB']
    else:
        xdb = os.path.join(xdir,'.xapers','xapian')

    if len(sys.argv) > 1:
        cmd = sys.argv[1]
    else:
        cmd = []

    cli = xapers.cli.UI(xdir)

    ########################################
    if cmd in ['add','a']:
        tags = None
        infile = None
        source = None
        prompt = False
        view = False

        argc = 2
        while True:
            if argc >= len(sys.argv):
                break
            elif '--source=' in sys.argv[argc]:
                source = sys.argv[argc].split('=',1)[1]
            elif '--file=' in sys.argv[argc]:
                infile = sys.argv[argc].split('=',1)[1]
            elif '--tags=' in sys.argv[argc]:
                tags = sys.argv[argc].split('=',1)[1].split(',')
            elif '--prompt' in sys.argv[argc]:
                prompt = True
            elif '--view' in sys.argv[argc]:
                view = True
            else:
                break
            argc += 1

        docid = cli.add(None, infile=infile, source=source, tags=tags, prompt=prompt)
        # dereference the cli object so that the database is flushed
        # FIXME: is there a better way to handle this?
        cli = None
        if view and docid:
            xapers.nci.UI(xdir, 'search', 'id:'+docid)

    ########################################
    elif cmd in ['update','u']:
        infile = None
        source = None

        argc = 2
        while True:
            if argc >= len(sys.argv):
                break
            elif '--source=' in sys.argv[argc]:
                source = sys.argv[argc].split('=',1)[1]
            elif '--file=' in sys.argv[argc]:
                infile = sys.argv[argc].split('=',1)[1]
            else:
                break
            argc += 1

        docid = sys.argv[argc]

        cli.add(docid, infile=infile, source=source)

    ########################################
    elif cmd == 'update-all':
        cli.update_all()

    ########################################
    elif cmd == 'delete':
        cli.delete(make_query_string(sys.argv[2:]))

    ########################################
    elif cmd == 'new':
        print >>sys.stderr, "not implemented."

    ########################################
    elif cmd in ['search','s']:
        if len(sys.argv) < 3:
            print >>sys.stderr, "Must specify a search term."
            sys.exit()

        oformat = 'simple'
        limit = 20

        argc = 2
        while True:
            if '--output=' in sys.argv[argc]:
                oformat = sys.argv[argc].split('=')[1]
            elif '--limit=' in sys.argv[argc]:
                limit = int(sys.argv[argc].split('=')[1])
            else:
                break
            argc += 1

        query = make_query_string(sys.argv[argc:])
        try:
            cli.search(query, oformat=oformat, limit=limit)
        except KeyboardInterrupt:
            sys.exit()

    ########################################
    elif cmd in ['bibtex','bib','b']:
        if len(sys.argv) < 3:
            print >>sys.stderr, "Must specify a search term."
            sys.exit()

        argc = 2
        query = make_query_string(sys.argv[argc:])
        try:
            cli.search(query, oformat='bibtex')
        except KeyboardInterrupt:
            sys.exit()

    ########################################
    elif cmd in ['view','v','show']:
        query = make_query_string(sys.argv[2:])
        if not query or query == '':
            query = 'tag:inbox'
        try:
            xapers.nci.UI(xdir, ['search', query])
        except KeyboardInterrupt:
            sys.exit()

    ########################################
    elif cmd in ['nci']:
        try:
            xapers.nci.UI(xdir, cmd=sys.argv[2:])
        except KeyboardInterrupt:
            sys.exit()

    ########################################
    elif cmd in ['tag','t']:
        add_tags = []
        remove_tags = []

        argc = 2
        for arg in sys.argv[argc:]:
            if arg == '--':
                argc += 1
                continue
            if arg[0] == '+':
                add_tags.append(arg[1:])
            elif arg[0] == '-':
                remove_tags.append(arg[1:])
            else:
                break
            argc += 1

        query = make_query_string(sys.argv[argc:])
        cli.tag(query, add_tags, remove_tags)

    ########################################
    elif cmd == 'dumpterms':
        query = make_query_string(sys.argv[2:])
        cli.dumpterms(query)

    ########################################
    elif cmd == 'count':
        query = make_query_string(sys.argv[2:])
        cli.count(query)

    ########################################
    elif cmd == 'export':
        outdir = sys.argv[2]
        query = make_query_string(sys.argv[3:])
        cli.export(outdir, query)

    ########################################
    elif cmd in ['source2bib','s2b']:
        string = sys.argv[2]
        import xapers.source
        bibtex, url = xapers.source.fetch_bibtex(string)
        if url:
            print >>sys.stderr, url
        try:
            print xapers.bibtex.Bibentry(bibtex).as_string()
        except:
            raise
            print >>sys.stderr, "Problem parsing bibtex.  Outputting raw bibtex..."
            print bibtex

    ########################################
    elif cmd in ['scandoc','sd']:
        infile = sys.argv[2]
        import xapers.source

        sources = xapers.source.scan_for_sources(infile)
        for ss in sources:
            print "%s" % (ss)

    ########################################
    elif cmd == 'version':
        print "Ha!"

    ########################################
    elif cmd in ['help','h']:
        usage()
        sys.exit(0)

    ########################################
    else:
        print >>sys.stderr, "unknown sub command '%s'." % cmd
        usage()
        sys.exit(1)
